File: 0232-implement-queue-using-stacks/0232-implement-queue-using-stacks.py
class MyQueue:
    def __init__(self):
        self.st1 = deque()
        self.st2 = deque()
        self.front = -1

    def push(self, x: int) -> None:
        if not self.st1: self.front = x
        self.st1.append(x)

    def pop(self) -> int:
        if self.st2: ans = self.st2.pop()
        else:
            while self.st1:
                self.st2.append(self.st1.pop())
            ans = self.st2.pop()
        return ans

    def peek(self) -> int:
        if self.st2: return self.st2[-1]
        else: return self.front

    def empty(self) -> bool:
        return len(self.st1) == 0 and len(self.st2) == 0


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: 0232-implement-queue-using-stacks/0232-implement-queue-using-stacks.cpp
class MyQueue {
public:
    stack<int> st1;
    stack<int> st2;
    int front;
    MyQueue() {
        front = -1;
    }
    
    void push(int x) {
        if(st1.empty()) front = x;
        st1.push(x);
    }
    
    int pop() {
        if(!st2.empty()) {
            int ans = st2.top();
            st2.pop();
            return ans;
        }
        else {
            while(!st1.empty()) {
                st2.push(st1.top());
                st1.pop();
            }
            int ans = st2.top();
            st2.pop();
            return ans;
        }
    }
    
    int peek() {
        if(!st2.empty()) return st2.top();
        else return front;
    }
    
    bool empty() {
        return st1.empty() && st2.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */